### PR TITLE
Queue behaviour

### PR DESCRIPTION
**Pull request by dorkamotorka**

### CODE REVIEW

**Comment by dorkamotorka** [comment]
goalCancelling test fixture is not working properly yet.

**Comment by tp4348** [comment]
So please review only to line 249.

File: test/test_goal_queueing.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+	// First goal
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear 
+	ASSERT_TRUE(qserv->isActive());
+	ASSERT_TRUE(got_goal);
+	ASSERT_EQ(3.0, received_goal->target_pose.pose.position.x);
+
+	// Cancelling the first goal - PITFALL
+	resumeExecuting();
```
Comment (rohbotics) [inline test/test_goal_queueing.cpp:157]: You should resume after you sent the cancel, that way the code in the callback has a chance to check for preempts and stuff again.

File: test/test_goal_queueing.cpp [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+	// "Second" goal
+	goal.target_pose.pose.position.x = 7.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear
+	ASSERT_TRUE(qserv->isActive());
+	ASSERT_TRUE(got_goal);
+	ASSERT_EQ(7.0, received_goal->target_pose.pose.position.x); 
+	
+	// Third goal
```
Comment (rohbotics) [inline test/test_goal_queueing.cpp]: This part seems reasonable

File: test/test_goal_queueing.cpp [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear
+	ASSERT_TRUE(qserv->isActive()); 
+	ASSERT_TRUE(got_goal);
+	ASSERT_FALSE(goal_preempted);
+	ASSERT_FALSE(next_goal_available);
+	ASSERT_EQ(13.0, received_goal->target_pose.pose.position.x);
+	finishExecuting(); // Finish the 3nd goal
+/*
+	- if another goal is received add it to the queue (DONE) 
+	- if the queue full, set the current goal as preempted - explicitly called! - so cancelling the current goal and start executing the next one
```
Comment (rohbotics) [inline test/test_goal_queueing.cpp]: We decided to change this behavior. If the queue is full, the last goal in the queue will be canceled and replaced with the new goal.

**Review by rohbotics — CHANGES_REQUESTED** [review]
Thanks, overall the tests look good, but all the sleeps make me nervous, because it means the tests are timing sensitive and might randomly fail. I ran the tests 10 times, and I had at least 3 random failures.

I think with more signalling condition variables we can significantly increase the reliability of the tests.
1) We should have a signal that goes off at the beginning of `executeCallback` then we can have a helper function that waits for that signal to go off instead of just doing a sleep.

2) resumeExecuting needs to wait for one loop of the executeCallback to complete so that the test variables actually have a chance to update.

Can you implement that?

**Comment by dorkamotorka** [comment]
Yes, I will work on it.

**Comment by dorkamotorka** [comment]
So despite removing the sleep functions and adding Helper function, unit testing succed in 7/10 cases. Before moving on to ASSERT-ing variables I want to make sure where am I missing things, that the test is so unreliable.

**Comment by tp4348** [comment]
The problem is that when cancelling goal in line 172, the goal message with ..x=7 should be "deleted/canceled" but it continues on in line 191 expected value is ..x=13 but it is 7. As you can see I tried with sleep and ros::spin. Previous commit is a bit more clear.

**Comment by tp4348** [comment]
In line 93 when I set execution=false (so that it updates in executeCallback while loop) catkin_make run_tests is not producing any result(I assume there is some blocking happening that I cannot seem to debug). 

In line 167 ASSERT should definitely work, because a new goal has been send to the Server.

In lines 170-189 there seem to be that the cancelGoal() is not working.

**Comment by dorkamotorka** [comment]
The test fixtures are working consistently but lagging a lot due to sleep() that cancelGoal and everything else required to work properly. Tests are time independent, sleep are there just so the threads have enough time to close.

File: test/test_goal_queueing.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	EXPECT_TRUE(got_goal);
+	EXPECT_EQ(3.0, received_goal->target_pose.pose.position.x);
+
+	cli->cancelGoal();
+	ros::Duration(1.0).sleep();
+	ros::spinOnce(); 
+	resumeExecuting();
+	updateExecuting();
+	sleepExecuting();
+	EXPECT_TRUE(goal_preempted);
```
Comment (rohbotics) [inline test/test_goal_queueing.cpp]: I tested this 10 times, this is the only line that failed (failed 3 times). Can you put a 0.5 second sleep before it as a quick fix?

**Review by rohbotics — CHANGES_REQUESTED** [review]

File: test/test_goal_queueing.cpp (hunk 4 above)
Comment (dorkamotorka) [reply test/test_goal_queueing.cpp]: Weird, never failed in my case. Yes, sure.

**Review by rohbotics — APPROVED** [review]

**Pull request merged**